packages/eslint-plugin/src/rules/consistent-type-imports.ts

```diff
@@ -0,0 +1,153 @@
+    },
+    schema: [
+      {
+        oneOf: [
+          {
+            enum: ['type-imports', 'no-type-imports'],
+          },
+          {
+            type: 'object',
+            properties: {
+              prefer: {
+                enum: ['type-imports', 'no-type-imports'],
+              },
+              disallowTypeAnnotations: {
+                type: 'boolean',
+              },
+            },
+            additionalProperties: false,
+          },
+        ],
+      },
+    ],
+    fixable: 'code',
+  },
+
+  defaultOptions: [
+    {
+      prefer: 'type-imports',
+      disallowTypeAnnotations: true,
+    },
+  ],
+
+  create(context, [option]) {
+    const prefer =
+      (typeof option === 'string' ? option : option.prefer) ?? 'type-imports';
+    const disallowTypeAnnotations =
+      typeof option === 'string'
+        ? true
+        : option.disallowTypeAnnotations !== false;
+    const sourceCode = context.getSourceCode();
+
+    const allValueImports: TSESTree.ImportDeclaration[] = [];
+    const referenceIdToDeclMap = new Map<
+      TSESTree.Identifier,
+      TSESTree.ImportDeclaration
+    >();
+    return {
+      ...(prefer === 'type-imports'
+        ? {
+            // prefer type imports
+            'ImportDeclaration[importKind=value]'(
+              node: TSESTree.ImportDeclaration,
+            ): void {
+              let used = false;
+              for (const specifier of node.specifiers) {
+                const id = specifier.local;
+                const variable = context
+                  .getScope()
+                  .variables.find(v => v.name === id.name)!;
+                for (const ref of variable.references) {
+                  if (ref.identifier !== id) {
+                    referenceIdToDeclMap.set(ref.identifier, node);
+                    used = true;
+                  }
+                }
+              }
+              if (used) {
+                allValueImports.push(node);
+              }
+            },
+            'TSTypeReference Identifier'(node: TSESTree.Identifier): void {
+              // Remove type reference ids
+              referenceIdToDeclMap.delete(node);
+            },
+            'Program:exit'(): void {
+              const usedAsValueImports = new Set(referenceIdToDeclMap.values());
+              for (const valueImport of allValueImports) {
+                if (usedAsValueImports.has(valueImport)) {
+                  continue;
+                }
+                context.report({
+                  node: valueImport,
+                  messageId: 'typeOverValue',
+                  fix(fixer) {
+                    // import type Foo from 'foo'
+                    //       ^^^^^ insert
+                    const importToken = sourceCode.getFirstToken(valueImport)!;
+                    return fixer.insertTextAfter(importToken, ' type');
+                  },
+                });
+              }
+            },
+          }
```

Comment:
This is a great start! But there are a few problems I can see:

- this code will incorrectly convert `import A, { B } from 'foo'` to `import type A, { B } from 'foo'`
    - type-only imports must _only_ have a default ***OR*** named imports, not both.
- this code explicitly ignores the case where you have a mixture of type and value imports
- the `TSTypeReference Identifier` selector is inaccurate, as it does not handle shadowing:
    ```ts
    import Type from 'foo';
    type T<Type> = Type; // this "Type" shadows the imported "Type", and will false positive
    ```

You should be able to lean much harder on the scope analyser here - as I built the scope analyser with some helpers to make it easier to inspect references.

I had a quick play, below is some code to help you. The difficult bit now will writing a resilient fixer:

```ts
{
  'ImportDeclaration[importKind = "value"]'(
    node: TSESTree.ImportDeclaration,
  ): void {
    const variables = context.getDeclaredVariables(node);
    const typeVariables: TSESLint.Scope.Variable[] = [];
    const valueVariables: TSESLint.Scope.Variable[] = [];
    for (const variable of variables) {
      const onlyHasTypeReferences = variable.references.every(ref => {
        if (ref.isTypeReference) {
          return true;
        }

        if (ref.isValueReference) {
          // `type T = typeof foo` will create a value reference because "foo" must be a value type
          // however this value reference is safe to use with type-only imports
          let parent = ref.identifier.parent;
          while (parent) {
            if (parent.type === AST_NODE_TYPES.TSTypeQuery) {
              return true;
            }
            // TSTypeQuery must have a TSESTree.EntityName as its child, so we can filter here and break early
            if (parent.type !== AST_NODE_TYPES.TSQualifiedName) {
              break;
            }
            parent = parent.parent;
          }
        }

        return false;
      });

      if (onlyHasTypeReferences) {
        typeVariables.push(variable);
      } else {
        valueVariables.push(variable);
      }
    }

    if (valueVariables.length === 0) {
      // import is all type-only, convert the entire import to `import type`
      // EXAMPLES:
      
      //   import { Type1, Type2 } from 'foo';
      // should be fixed to:
      //   import type { Type1, Type2 } from 'foo';

      //   import Type from 'foo';
      // should be fixed to:
      //   import type Type from 'foo';

      //   import * as Type from 'foo';
      // should be fixed to:
      //   import type * as Type from 'foo';

      // !!! NOTE: must take special care in this case:
      //   import Default, { Named } from 'foo';
      // should be fixed to:
      //   import type Default from 'foo';
      //   import type { Named } from 'foo';

      // context.report({ message: 'All imports in the declaration are only used as types (or some error like that)' })
    } else {
      // we have a mixed type/value import, so we need to split them out into multiple exports
      // EXAMPLES:

      //   import { Value, Type } from 'foo';
      // should be fixed to:
      //   import type { Type } from 'foo';
      //   import { Value } from 'foo';

      //   import Type, { Value } from 'foo';
      // should be fixed to:
      //   import type Type from 'foo';
      //   import { Value } from 'foo';

      //   import Value, { Type } from 'foo';
      // should be fixed to:
      //   import type { Type } from 'foo';
      //   import Value from 'foo';

      // !!! NOTE: must take special care in this case:
      //   import Type1, { Type2, Value } from 'foo';
      // should be fixed to:
      //   import type Type1 from 'foo';
      //   import type { Type2 } from 'foo';
      //   import { Value } from 'foo';
      
      // context.report({ message: 'Imports "A", "B" and "C" are only used as types (or some error like that)' })
    }
  },
}
```